cmd/kots/cli/template.go

```diff
@@ -0,0 +1,339 @@
+	return &templateReplSession{
+		b: builder,
+		l: log,
+	}
+}
+
+func runInteractive(b *template.Builder, log *logger.CLILogger) error {
+	repl := createReplSession(b, log)
+	return repl.run()
+}
+
+func (r *templateReplSession) run() error {
+	rl, err := readline.NewEx(&readline.Config{
+		Prompt:            "> ",
+		InterruptPrompt:   "^C",
+		EOFPrompt:         "exit",
+		HistorySearchFold: true,
+		Stdin:             os.Stdin,
+		Stdout:            os.Stdout,
+		Stderr:            os.Stderr,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize console")
+	}
+	defer rl.Close()
+
+	for {
+		line, err := rl.Readline()
+		if err == readline.ErrInterrupt {
+			if len(line) == 0 {
+				break
+			} else {
+				continue
+			}
+		} else if err == io.EOF {
+			break
+		}
+
+		out, exit, err := r.handle(line)
+		if exit {
+			break
+		}
+		r.l.Info(out)
+	}
+
+	return nil
+}
+
+func (r *templateReplSession) handle(line string) (string, bool, error) {
+	switch {
+	case strings.TrimSpace(line) == "exit":
+		return "", true, nil
+	case strings.TrimSpace(line) == "help":
+		return r.help(), false, nil
+	default:
+		rendered, err := r.b.String(line)
+		return rendered, false, err
+	}
+}
+
+func (r *templateReplSession) help() string {
+	return `
+	Go to https://docs.replicated.com/reference/template-functions-about for a list of available template functions.
+	Available commands:
+ 		help - display this help message
+  		exit - exit the interactive console
+`
+}
+
+func pullAndRender(appSlug string, licensePath string, configPath string) error {
+	tempDir, err := os.MkdirTemp("", "kots-template")
+	if err != nil {
+		return errors.Wrap(err, "failed to create temp directory to render templates")
+	}
+	defer os.RemoveAll(tempDir)
+
+	pullOptions := pull.PullOptions{
+		RootDir:             tempDir,
+		AppSlug:             appSlug,
+		LicenseFile:         ExpandDir(licensePath),
+		ConfigFile:          ExpandDir(configPath),
+		Silent:              true,
+		ExcludeAdminConsole: true,
+	}
+
+	upstream := pull.RewriteUpstream(appSlug)
+	_, err = pull.Pull(upstream, pullOptions)
+
+	if err != nil && err != pull.ErrConfigNeeded {
+		return errors.Wrap(err, "failed to pull upstream")
+	}
+
+	// iterate over kotsKinds directory in tempDir and print all YAML contents
+	kotsKindsDir := filepath.Join(tempDir, "kotsKinds")
+	files, err := os.ReadDir(kotsKindsDir)
+	if err != nil {
+		return errors.Wrap(err, "failed to read kotsKinds directory")
+	}
+
+	manifetsToRender := make(map[string]string)
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		content, err := os.ReadFile(filepath.Join(kotsKindsDir, file.Name()))
+		if err != nil {
+			return errors.Wrap(err, "failed to read file")
+		}
+		manifetsToRender[file.Name()] = string(content)
+	}
+	for k, m := range manifetsToRender {
+		fmt.Println("---")
+		fmt.Printf("# Source: %s\n", k)
+		fmt.Println(m)
+	}
+
+	// also render helm charts with helm template if any
```

Comment:
the `rendered` directory should already have the rendered helm templates, plain k8s manifests, and additional resources created by kustomize (pull secrets). i think we may be able to instead just iterate and print the files in that directory?

the kotskinds will not be in `rendered`, so i think we can keep what you have above this to show them to the user

Reply:
Many thanks Craig! I've included `rendered` directory excluding `kotsadm-` files. However I think we should also run `helm template` on `helm/` directory for HelmChart v2. What do you think?

Reply:
the `rendered` directory will already contain the `helm template` output in an `all.yaml` file for each HelmChart, so i think we can just iterate over all of the files there, including the `kotsadm-` files since those will ultimately be deployed along with everything else

Reply:
gotcha! I've removed the helm template part. Please let me know how it goes!